### PR TITLE
[MC/DC][Coverage] Introduce "Bitmap Bias" for continuous mode

### PR DESCRIPTION
**Pull request by chapuni**
`counter_bias` is incompatible to Bitmap. The distance between Counters and Bitmap is different between on-memory sections and profraw image.

Reference to `__llvm_profile_bitmap_bias` is generated only if `-fcoverge-mcdc` `-runtime-counter-relocation` are specified. The current implementation rejected their options.

```
Runtime counter relocation is presently not supported for MC/DC bitmaps
```

### CODE REVIEW

**Comment by petrhosek** [comment]
This is something we discussed internally when reviewing the MC/DC changes. There are two different design directions we can take here: use a separate bias for the bimap (as done in this change), or use the same bias for both counters and bitmap.

The downside of using separate biases is that it can have significant negative performance impact. We intentionally load the counter bias at the beginning of the function, with the motivation being that unless we're under register pressure, the register allocator will use the same register throughout the entire function, reducing the number of loads. This has been the case in practice from what we've observed. Using a second bias is likely going to increase register pressure, especially on register-starved architectures like x86, and require more frequent loads for the bias.

A more efficient alternative would be to use a single bias for both counters and bitmap. That would require the counters and the bitmap section being at a fixed offset, ideally right next to each other. In ELF, we might be able to achieve this using `SHF_LINK_ORDER` but we haven't tested this to see if it's really feasible. We also don't know if there's a way to represent this in COFF and Mach-O, that's something we would need to look into.

**Comment by chapuni** [comment]
@petrhosek Thanks for the comment.

I don't think this would introduce serious performance regression, and I don't think coverage-instrumented codes would be sensitive of performance so much.

I know it'd be better to use the single bias offset between counters and bitmaps. We might do in the future. This is suboptimal but practical, I suppose.

It'd be the best if we could map and override the profraw into the existing profc and profbm, as Darwin is doing.
For it, I guess we would modify the layout of profc and profbm as aligned to phys page.
(Darwin's impl doesn't use contiguous mapping but individual mappings to profc and profbm)

I'll propose the bitmap for coverage in the near future. It will not require counters.
At the moment, I think 2nd bias offset would be practical for supporting mcdc with continuous mode.

**Comment by ornata** [comment]
Do we have an expected % performance impact?

If perf is critical, could it be possible to only handle ELF for now, and support COFF + Mach-O later?  My intuition says that most folks that care about MC/DC are going to primarily be working with ELF.

**Comment by chapuni** [comment]
Ping. This will affect only if mcdc coverage is enabled.

File: compiler-rt/lib/profile/InstrProfilingFile.c [hunk 1]
```diff
@@ -260,7 +282,17 @@ static int mmapForContinuousMode(uint64_t CurrentFileOffset, FILE *File) {
...
+   * will fail with EINVAL. */
+  if (NumBitmapBytes == 0)
+    return 0;
+
+  /* Update profbm_bias. */
+  uint64_t FileOffsetToBitmap =
+      FileOffsetToCounters + CountersSize + PaddingBytesAfterCounters;
+  /* Update the profile fields based on the current mapping. */
+  INSTR_PROF_PROFILE_BITMAP_BIAS_VAR =
+      (intptr_t)Profile - (uintptr_t)BitmapBegin + FileOffsetToBitmap;
```
Comment (ornata) [inline compiler-rt/lib/profile/InstrProfilingFile.c:295]: should profile be uintptr_t?

File: llvm/lib/Transforms/Instrumentation/InstrProfiling.cpp [hunk 2]
```diff
@@ -1014,18 +1014,22 @@ CallInst *InstrLowerer::getRMWOrCall(Value *Addr, Value *Val) {
...
-              "bitmaps."),
-        DS_Warning));
-  }
+  if (!isRuntimeCounterRelocationEnabled())
+    return Bitmaps;
 
-  return Bitmaps;
+  // Put BiasLI onto the entry block.
+  Type *Int64Ty = Type::getInt64Ty(M.getContext());
+  Function *Fn = I->getParent()->getParent();
```
Comment (ornata) [inline llvm/lib/Transforms/Instrumentation/InstrProfiling.cpp:1022]: nit: getFunction() ?

**Review by ornata — APPROVED** [review]
This only impacts MC/DC with continuous mode enabled, so I think it's fine to merge this.

I think that we should try to make this more performant though.

File: compiler-rt/lib/profile/InstrProfilingFile.c (hunk 1 above)
Comment (chapuni) [reply compiler-rt/lib/profile/InstrProfilingFile.c:295]: Yes, it should be unaware of wrapping-around. It was copypasto from `INSTR_PROF_PROFILE_COUNTER_BIAS_VAR`.

File: llvm/lib/Transforms/Instrumentation/InstrProfiling.cpp (hunk 2 above)
Comment (chapuni) [reply llvm/lib/Transforms/Instrumentation/InstrProfiling.cpp:1022]: I can see more `getParent()->getParent()` than `getFunction()` in this file :)

**Comment by chapuni** [comment]
I am certain that this change won't affect current valid configurations. I will merge this tomorrow, several hours after, unless seeing strong objections.

**Pull request merged**